apps/src/netsim/NetSimRouterNode.js

```diff
@@ -292,11 +358,140 @@ NetSimRouterNode.prototype.buildRow_ = function () {
 };
 
 /**
+ * Load state from remoteRow into local model, then notify anything observing
+ * us that we've changed.
+ * @param {routerRow} remoteRow
+ * @private
+ */
+NetSimRouterNode.prototype.onMyStateChange_ = function (remoteRow) {
+  this.bandwidth = deserializeNumber(remoteRow.bandwidth);
+  this.dnsMode = remoteRow.dnsMode;
+  this.dnsNodeID = remoteRow.dnsNodeID;
+  this.stateChange.notifyObservers(this);
+};
+
+/**
  * Ticks heartbeat, telling the network that router is in use.
  * @param {RunLoop.Clock} clock
  */
 NetSimRouterNode.prototype.tick = function (clock) {
+  this.simulationTime_ = clock.time;
   this.heartbeat_.tick(clock);
+  this.tickRouting_.call(this, clock);
+  if (this.dnsMode === DnsMode.AUTOMATIC) {
+    this.tickAutoDns_(clock);
+  }
+};
+
+/**
+ * @param {RunLoop.Clock} clock
+ * @private
+ */
+NetSimRouterNode.prototype.tickRouting_ = function (clock) {
+  if (this.isRouterProcessing_) {
+    return;
+  }
+
+  // Process requests that are at or past their scheduled time.
+  var dueForRouting = this.getAndUnscheduleMessagesDueForRouting(clock);
+  if (dueForRouting.length > 0) {
+    this.isRouterProcessing_ = true;
+    this.routeMessages_(dueForRouting, function () {
+      this.isRouterProcessing_ = false;
+    }.bind(this));
+  }
+};
+
+/**
+ * Destructive operation: Removes schedule entries that are at or past their
+ * scheduled time, generates the message objects for those schedule entries,
+ * and returns them.
+ * @param {RunLoop.clock} clock
+ * @returns {NetSimMessage[]}
+ */
+NetSimRouterNode.prototype.getAndUnscheduleMessagesDueForRouting = function (clock) {
+  var dueForRouting = [];
+  this.localRoutingSchedule_ = this.localRoutingSchedule_.filter(function (entry) {
+    if (clock.time >= entry.time) {
+      dueForRouting.push(new NetSimMessage(this.shard_, entry.row));
+      return false;
+    }
+    return true;
+  }.bind(this));
```

Comment:
Interesting way of doing this. I can't decide if I like it or not. Another way you could approach it is to have two arrays dueForRouting and notDueForRouting, then forEach localRoutingSchedule put it in one or the other, then at the end set localRoutingSchedule_ = notDueForRouting. Might be a little clearer exactly what's happening. Up to you which you prefer.